components/api_request_helper/api_request_helper.cc

```diff
@@ -335,106 +315,189 @@ void APIRequestHelper::URLLoaderHandler::send_sse_data_for_testing(
-  auto error_code = result.error_code();
-  auto final_url = result.final_url();
+  VLOG(1) << "[[" << __func__ << "]]"
+          << " Response received\n";
+  DCHECK(result_callback_);
+  // This shouldn't be called on a request with multiple decoding operations,
+  // otherwise we need to modify this to use data_chunk_parsed_callback_.
+  DCHECK_EQ(current_decoding_operation_count_, 0);
+  APIRequestResult result = ToAPIRequestResult(std::move(url_loader_));
 
   if (!response_body) {
-    std::move(callback).Run(APIRequestResult(response_code, "", base::Value(),
-                                             std::move(headers), error_code,
-                                             final_url));
+    std::move(result_callback_).Run(std::move(result));
     return;
   }
   auto& raw_body = *response_body;
   if (conversion_callback) {
     auto converted_body = std::move(conversion_callback).Run(raw_body);
     if (!converted_body) {
-      std::move(callback).Run(APIRequestResult(
-          422, "", base::Value(), std::move(headers), error_code, final_url));
+      result.response_code_ = 422;
+      std::move(result_callback_).Run(std::move(result));
       return;
     }
     raw_body = converted_body.value();
   }
 
-  api_request_helper_->data_decoder_->ParseJson(
-      raw_body,
-      base::BindOnce(&OnParseJsonIsolated, response_code, std::move(headers),
-                     error_code, final_url, std::move(callback)));
+  GetDataDecoder()->ParseJson(
+      std::move(raw_body),
+      base::BindOnce(&APIRequestHelper::URLLoaderHandler::OnParseJsonResponse,
+                     GetWeakPtr(), std::move(result)));
+}
 
-  api_request_helper_->Cancel(this);
+void APIRequestHelper::URLLoaderHandler::OnParseJsonResponse(
+    APIRequestResult result,
+    data_decoder::DataDecoder::ValueOrError result_value) {
+  // TODO(petemill): Simplify by combining OnParseJsonResponse with the Json
+  // response handler in ParseSSE.
+  if (!result_value.has_value()) {
+    VLOG(1) << "Response validation error:" << result_value.error();
+    std::move(result_callback_).Run(std::move(result));
+    return;
+  }
+  if (!result_value.value().is_dict() && !result_value.value().is_list()) {
+    VLOG(1) << "Response validation error: Invalid top-level type";
+    std::move(result_callback_).Run(std::move(result));
+    return;
+  }
+  std::string safe_json;
+  if (!base::JSONWriter::Write(result_value.value(), &safe_json)) {
+    VLOG(1) << "Response validation error: Encoding error";
+    std::move(result_callback_).Run(std::move(result));
+    return;
+  }
+  VLOG(2) << "Reponse validation successful";
+  result.body_ = std::move(safe_json);
+  result.value_body_ = std::move(result_value.value());
+  std::move(result_callback_).Run(std::move(result));
+}
+
+void APIRequestHelper::URLLoaderHandler::MaybeSendResult() {
+  // Verify that counting hasn't gone wrong - it should never be less than 0.
+  DCHECK_LT(-1, current_decoding_operation_count_);
```

Comment:
nit: `DCHECK_LE(0, current_decoding_operation_count_)` would be a bit more precise

Reply:
changed it